js/src/elements/form-behavior.ts

```diff
@@ -0,0 +1,744 @@
+	const group = input.closest<HTMLElement>('.field');
+	if (!group) {
+		throw new Error(
+			`${input.nodeName} name="${input.name}" id="${input.id}" is not within a .field`
+		);
+	}
+	return group;
+}
+
+function getFieldBody(input: HTMLValueElement) {
+	const body = input.closest('.field-body');
+	if (!body) {
+		throw new Error(
+			`${input.nodeName} name="${input.name}" id="${input.id}" is not within a .field-body`
+		);
+	}
+	return body;
+}
+
+function createErrorNote(input: HTMLValueElement) {
+	const note = document.createElement('p');
+	note.id = generateElementId();
+	input.setAttribute(
+		'aria-describedby',
+		`${note.id} ${input.getAttribute('aria-describedby') || ''}`
+	);
+	note.classList.add('field-error');
+	getFieldBody(input).after(note);
+	return note;
+}
+
+function setValidationMessage(element: HTMLValueElement, message: string) {
+	const group = getField(element);
+	const note = group.querySelector('.field-error') || createErrorNote(element);
+	note.textContent = message;
+}
+
+function canValidate(
+	target: EventTarget | null,
+	form: HTMLFormElement | null
+): target is HTMLValueElement {
+	return isValueElement(target, form) && (target as HTMLValueElement).type !== 'hidden';
+}
+
+export function navigateAfterSubmit(href: string, navigationStep: string | null) {
+	switch (navigationStep) {
+		case null:
+			// do nothing.
+			break;
+		case 'follow':
+			if (href) {
+				location.href = href;
+			}
+			break;
+		case 'replace':
+			if (href) {
+				location.replace(href);
+			}
+			break;
+		case 'reload':
+			location.reload();
+			break;
+		default:
+			throw new Error(`Unexpected navigation attribute value.`);
+	}
+}
+
+/* Custom elements are not included in HTMLFormControlsCollection so they can't be retrieved by form.elements */
+export function collectCustomElementsByName(form: HTMLFormElement): Element[] {
+	// Compare FormData with form.elements to find missing elements.
+	const formData = Object.fromEntries(new FormData(form));
+	const formElements = Object.keys(form.elements);
+	const customElementList: Element[] = [];
+	const customElements = Object.keys(formData).filter(el => formElements.indexOf(el) === -1);
+	customElements.forEach(name => {
+		const element = form.querySelector(`[name="${name}"]`);
+		if (element) {
+			customElementList.push(element);
+		}
+	});
+	return customElementList;
+}
+
+function clearInputErrorBorder(input: HTMLValueElement) {
+	const isTagSelector = input.classList.contains('tag-input');
+	if (isTagSelector) {
+		input.nextElementSibling?.classList.remove('border-color-danger');
+	}
+	input.classList.remove(`${input.localName}-danger`);
+}
+
+function handleSubmitButtonAction(event: Event) {
+	const submitter = (event as SubmitEvent).submitter;
+	const formAction =
+		submitter instanceof HTMLButtonElement && submitter.formAction ? submitter.formAction : null;
+	const form = event.currentTarget as HTMLFormElement;
+
+	if (formAction) {
+		form.action = formAction;
```

Comment:
Thinking it would be best to return the formAction rather than overwriting the form's action.

Reply:
Updated